src/Orleans/Serialization/BinaryTokenStreamReader.cs

```diff
@@ -115,52 +126,52 @@ private ArraySegment<byte> CheckLength(int n, out bool safeToUse)
                     CurrentPosition, n, totalLength));
             }
 
-            if (currentSegmentIndex >= buffers.Count)
+            if (currentSegmentIndex >= buffersCount)
             {
                 throw new SerializationException(
                     String.Format("Attempt to read past buffers.Count: currentSegmentIndex={0}, buffers.Count={1}.", currentSegmentIndex, buffers.Count));
             }
 
-            if (currentOffset == currentSegment.Offset + currentSegment.Count)
+            if (currentOffset == currentSegmentOffset + currentSegmentCount)
             {
                 StartNextSegment();
             }
 
-            if (currentOffset + n <= currentSegment.Offset + currentSegment.Count)
+            var nextOffset = currentOffset + n;
+            if (nextOffset <= currentSegmentOffset + currentSegmentCount)
             {
-                var result = new ArraySegment<byte>(currentBuffer, currentOffset, n);
-                currentOffset += n;
-                if (currentOffset >= currentSegment.Offset + currentSegment.Count)
```

Comment:
are you sure? Without going into too much depth (you might have already checked and can guide me), they don't look equal. This was checking if `currentOffset` (which is now `+n` actually) is *greater* or equal.